### ISSUE TITLE
Segmentation fault in ext/maxminddb.c:418

### ISSUE THREAD

**Issue by thias** [maxmind/MaxMind-DB-Reader-php#20]
Hi,

My developers get a segmentation fault when trying to integrate the maxminddb module into their testing suite. Apparently some simple accesses work, but when doing some less usual stuff (extending a class and overriding its constructor, I think) it breaks. This is with the 1.0.0 code (you can see the exact hash in the debuginfo file path) and PHP 5.5.19 with GeoIP2-Country.mmdb.

Let me know if you need more information.

```
Core was generated by `php bin/phpspec run'.
Program terminated with signal 11, Segmentation fault.
#0  zend_hash_destroy (ht=0x0) at /usr/src/debug/php-5.5.19/Zend/zend_hash.c:555
555		p = ht->pListHead;
#0  zend_hash_destroy (ht=0x0) at /usr/src/debug/php-5.5.19/Zend/zend_hash.c:555
#1  0x00007fb7d496b574 in maxminddb_free_storage (object=0x2729ea8) at /usr/src/debug/MaxMind-DB-Reader-php-cbd695a5309c15249ef5944997cfb71f2a8a7963/ext/maxminddb.c:418
#2  0x00000000005e4aac in zend_objects_store_del_ref_by_handle_ex (handle=3100, handlers=<value optimized out>) at /usr/src/debug/php-5.5.19/Zend/zend_objects_API.c:226
#3  0x00000000005db4b8 in gc_collect_cycles () at /usr/src/debug/php-5.5.19/Zend/zend_gc.c:834
#4  0x00000000005db8b4 in gc_zval_possible_root (zv=<value optimized out>) at /usr/src/debug/php-5.5.19/Zend/zend_gc.c:166
#5  0x0000000000641b9e in gc_zval_check_possible_root (execute_data=0x7fb7e6b63898) at /usr/src/debug/php-5.5.19/Zend/zend_gc.h:183
#6  i_zval_ptr_dtor (execute_data=0x7fb7e6b63898) at /usr/src/debug/php-5.5.19/Zend/zend_execute.h:90
#7  ZEND_ASSIGN_SPEC_CV_VAR_HANDLER (execute_data=0x7fb7e6b63898) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:36980
#8  0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b63898) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#9  0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b63898) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#10 0x000000000063dd50 in zend_do_fcall_common_helper_SPEC (execute_data=0x7fb7e6b63380) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:584
#11 0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b63380) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#12 0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b63380) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#13 0x000000000063dd50 in zend_do_fcall_common_helper_SPEC (execute_data=0x7fb7e6b62c48) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:584
#14 0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b62c48) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#15 0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b62c48) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#16 0x000000000063dd50 in zend_do_fcall_common_helper_SPEC (execute_data=0x7fb7e6b62500) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:584
#17 0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b62500) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#18 0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b62500) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#19 0x000000000063dd50 in zend_do_fcall_common_helper_SPEC (execute_data=0x7fb7e6b62138) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:584
#20 0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b62138) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#21 0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b62138) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#22 0x000000000063dd50 in zend_do_fcall_common_helper_SPEC (execute_data=0x7fb7e6b61a60) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:584
#23 0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b61a60) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#24 0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b61a60) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#25 0x000000000063dd50 in zend_do_fcall_common_helper_SPEC (execute_data=0x7fb7e6b61718) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:584
#26 0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b61718) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#27 0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b61718) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#28 0x000000000063dd50 in zend_do_fcall_common_helper_SPEC (execute_data=0x7fb7e6b61450) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:584
#29 0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b61450) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#30 0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b61450) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#31 0x000000000063dd50 in zend_do_fcall_common_helper_SPEC (execute_data=0x7fb7e6b61328) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:584
#32 0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b61328) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#33 0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b61328) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#34 0x00000000005ae4dc in zend_call_function (fci=0x7fffd1547660, fci_cache=<value optimized out>) at /usr/src/debug/php-5.5.19/Zend/zend_execute_API.c:937
#35 0x00000000004f11a8 in zif_array_map (ht=<value optimized out>, return_value=0x33a1720, return_value_ptr=<value optimized out>, this_ptr=<value optimized out>, return_value_used=<value optimized out>) at /usr/src/debug/php-5.5.19/ext/standard/array.c:4371
#36 0x00000000005ab319 in dtrace_execute_internal (execute_data_ptr=<value optimized out>, fci=<value optimized out>, return_value_used=<value optimized out>) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:97
#37 0x000000000063d662 in zend_do_fcall_common_helper_SPEC (execute_data=<value optimized out>) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:552
#38 0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b61060) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#39 0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b61060) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#40 0x000000000063dd50 in zend_do_fcall_common_helper_SPEC (execute_data=0x7fb7e6b60dc0) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:584
#41 0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b60dc0) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#42 0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b60dc0) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#43 0x000000000063dd50 in zend_do_fcall_common_helper_SPEC (execute_data=0x7fb7e6b60c70) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:584
#44 0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b60c70) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#45 0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b60c70) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#46 0x00000000005ae4dc in zend_call_function (fci=0x7fffd1547d90, fci_cache=<value optimized out>) at /usr/src/debug/php-5.5.19/Zend/zend_execute_API.c:937
#47 0x00000000004fe7e7 in zif_call_user_func_array (ht=<value optimized out>, return_value=0x33abce0, return_value_ptr=<value optimized out>, this_ptr=<value optimized out>, return_value_used=<value optimized out>) at /usr/src/debug/php-5.5.19/ext/standard/basic_functions.c:4806
#48 0x00000000005ab319 in dtrace_execute_internal (execute_data_ptr=<value optimized out>, fci=<value optimized out>, return_value_used=<value optimized out>) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:97
#49 0x000000000063d662 in zend_do_fcall_common_helper_SPEC (execute_data=<value optimized out>) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:552
#50 0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b608e0) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#51 0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b608e0) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#52 0x00000000005ae4dc in zend_call_function (fci=0x7fffd1548170, fci_cache=<value optimized out>) at /usr/src/debug/php-5.5.19/Zend/zend_execute_API.c:937
#53 0x00000000005d36f7 in zend_call_method (object_pp=0x7fffd1548298, obj_ce=<value optimized out>, fn_proxy=0x21a4908, function_name=0x6cc3f6 "__call", function_name_len=<value optimized out>, retval_ptr_ptr=0x7fffd15482a8, param_count=2, arg1=0x33ab198, arg2=0x33abe88) at /usr/src/debug/php-5.5.19/Zend/zend_interfaces.c:97
#54 0x00000000005e23ed in zend_std_call_user_call (ht=<value optimized out>, return_value=0x33b3a30, return_value_ptr=<value optimized out>, this_ptr=0x30992b0, return_value_used=<value optimized out>) at /usr/src/debug/php-5.5.19/Zend/zend_object_handlers.c:899
#55 0x00000000005ab319 in dtrace_execute_internal (execute_data_ptr=<value optimized out>, fci=<value optimized out>, return_value_used=<value optimized out>) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:97
#56 0x000000000063d662 in zend_do_fcall_common_helper_SPEC (execute_data=<value optimized out>) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:552
#57 0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b60768) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#58 0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b60768) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#59 0x00000000005ae4dc in zend_call_function (fci=0x7fffd1548620, fci_cache=<value optimized out>) at /usr/src/debug/php-5.5.19/Zend/zend_execute_API.c:937
#60 0x00000000004b96b6 in zim_reflection_method_invokeArgs (ht=<value optimized out>, return_value=0x33a9aa0, return_value_ptr=<value optimized out>, this_ptr=<value optimized out>, return_value_used=<value optimized out>) at /usr/src/debug/php-5.5.19/ext/reflection/php_reflection.c:3018
#61 0x00000000005ab319 in dtrace_execute_internal (execute_data_ptr=<value optimized out>, fci=<value optimized out>, return_value_used=<value optimized out>) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:97
#62 0x000000000063d662 in zend_do_fcall_common_helper_SPEC (execute_data=<value optimized out>) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:552
#63 0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b604b0) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#64 0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b604b0) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#65 0x000000000063dd50 in zend_do_fcall_common_helper_SPEC (execute_data=0x7fb7e6b5fec8) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:584
#66 0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b5fec8) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#67 0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b5fec8) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#68 0x000000000063dd50 in zend_do_fcall_common_helper_SPEC (execute_data=0x7fb7e6b5f6b0) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:584
#69 0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b5f6b0) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#70 0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b5f6b0) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#71 0x000000000063dd50 in zend_do_fcall_common_helper_SPEC (execute_data=0x7fb7e6b5f1b8) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:584
#72 0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b5f1b8) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#73 0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b5f1b8) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#74 0x000000000063dd50 in zend_do_fcall_common_helper_SPEC (execute_data=0x7fb7e6b5ec38) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:584
#75 0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b5ec38) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#76 0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b5ec38) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#77 0x000000000063dd50 in zend_do_fcall_common_helper_SPEC (execute_data=0x7fb7e6b5e640) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:584
#78 0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b5e640) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#79 0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b5e640) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#80 0x000000000063dd50 in zend_do_fcall_common_helper_SPEC (execute_data=0x7fb7e6b5dfb8) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:584
#81 0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b5dfb8) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#82 0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b5dfb8) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#83 0x000000000063dd50 in zend_do_fcall_common_helper_SPEC (execute_data=0x7fb7e6b5d768) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:584
#84 0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b5d768) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#85 0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b5d768) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#86 0x000000000063dd50 in zend_do_fcall_common_helper_SPEC (execute_data=0x7fb7e6b5d0d0) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:584
#87 0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b5d0d0) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#88 0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b5d0d0) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#89 0x000000000063dd50 in zend_do_fcall_common_helper_SPEC (execute_data=0x7fb7e6b5ca18) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:584
#90 0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b5ca18) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#91 0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b5ca18) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#92 0x000000000063dd50 in zend_do_fcall_common_helper_SPEC (execute_data=0x7fb7e6b5c480) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:584
#93 0x000000000062f138 in execute_ex (execute_data=0x7fb7e6b5c480) at /usr/src/debug/php-5.5.19/Zend/zend_vm_execute.h:363
#94 0x00000000005ab44e in dtrace_execute_ex (execute_data=0x7fb7e6b5c480) at /usr/src/debug/php-5.5.19/Zend/zend_dtrace.c:73
#95 0x00000000005ba44c in zend_execute_scripts (type=8, retval=0x0, file_count=3) at /usr/src/debug/php-5.5.19/Zend/zend.c:1330
#96 0x000000000055a8b0 in php_execute_script (primary_file=0x7fffd154be70) at /usr/src/debug/php-5.5.19/main/main.c:2506
#97 0x000000000066bbe3 in do_cli (argc=3, argv=0xebdcc0) at /usr/src/debug/php-5.5.19/sapi/cli/php_cli.c:994
#98 0x000000000066c3e8 in main (argc=3, argv=0xebdcc0) at /usr/src/debug/php-5.5.19/sapi/cli/php_cli.c:1378
```

**Comment by Epilgrim** [comment]
You can find a way to reproduce this error here:
https://github.com/Epilgrim/geoip2_segfault_test

**Comment by oschwald** [comment]
Thanks for providing the test case. We'll look into this.

**Comment by oschwald** [comment]
Could you test #21 to see if the fix works for you?

**Comment by Epilgrim** [comment]
it seems it does. I was able to run the tests successfully against this patched version

**Comment by oschwald** [comment]
Thanks. We released 1.0.2 with this fix.
